news/_posts/2019-12-25-prestashop-in-2019-and-beyond-part-3-the-future-architecture.md

```diff
@@ -0,0 +1,269 @@
+
+The same can be said about in-memory static cache, which is another classic performance hack that ends up producing hair-pulling bugs. This technique is used to store data in cache across multiple calls to the same method, **even if those method calls are performed on different instances of the same class**. More often than not, this points to a design flaw and can be avoided by either refactoring or injecting cache providers (which also provide more convenient and powerful storage and invalidation options).
+
+#### No clearly-defined API / Everything is public
+
+One of the biggest problems in PrestaShop is that since it has no clearly-defined public API, everything is _de facto_ considered to belong to it. 
+
+The meanest part of this issue is the overrides system, which allows to redefine any public and protected method in any legacy class. The problem with it is that by violating the [Open/Closed principle](https://en.wikipedia.org/wiki/Open%E2%80%93closed_principle), it prevents the Core from evolving and from ensuring that its internal state is correct at all times.
+
+Classes that cannot be overridden also suffer from this problem. Even if they were originally meant for internal use, the _statu quo_ indicates that as soon as they exist and are released, somebody could be using them, and therefore, any change in their behavior or interface is considered a breaking change.
+
+This problem can be solved by clearly stating what belongs to the public API and what does not.
+
+### Driving forces
+
+Designing a new architecture is not just about solving problems. It also has to answer to a certain philosophy, a way of seeing things that drives our design choices and settles our priorities. What's important for us and non-negotiable? What are the trade-offs that we are willing to accept?
+
+#### It has to be coherent
+
+Coherence should be of utmost importance in a new architecture. Not only there should be one and only one way of performed a given task, but it should also be intuitive both in its implementation as well as its discoverability (i.e. avoid the classic question of _"Where's the class that handles that feature again?"_).
+
+In addition, a well-designed architecture should be _internally_ coherent and resilient to errors. The Core should therefore enforce coherency in its internal state and reject any possibility of incoherent input, by design. This means, among other things, strict typing and typed exceptions.
```

Comment:
You mentioned typed exceptions 3 times so far, I think an external source would be worth it for people who have not seen such a thing before to understand the concept

Reply:
I can't find a good article about that subject that I can agree with 😅

...I guess I will have to write my own!